### DISCUSSION TITLE
TS011F 2 Plug-Version - how to integrate?

### DISCUSSION THREAD

**Discussion by thebang2** [Koenkk/zigbee2mqtt#18500] [Devices]
Hi everybody, this is my first experience with an unknown Zigbee-Device. I´ve bought an "Immax NEO LITE Smart two socket".

https://zigbee.blakadder.com/Immax_07752L.html

The device is recognized in Z2M as "TuYa TS011F_plug_1", but I think this is the 1-plug version. For the first socket, everything is working and can be controlled:

`
2023-08-02 17:27:20Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,0]}' from endpoint 1 with groupID 0
2023-08-02 17:27:20Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'haElectricalMeasurement', data '{"activePower":0,"rmsCurrent":0,"rmsVoltage":231}' from endpoint 1 with groupID 0
2023-08-02 17:28:02Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genBasic', data '{"65506":54,"65508":0,"appVersion":77}' from endpoint 1 with groupID 0
2023-08-02 17:29:20Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
2023-08-02 17:29:20Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
2023-08-02 17:29:27Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'haElectricalMeasurement', data '{"rmsCurrent":67}' from endpoint 1 with groupID 0
2023-08-02 17:29:27Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'haElectricalMeasurement', data '{"activePower":22}' from endpoint 1 with groupID 0
2023-08-02 17:29:32Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'haElectricalMeasurement', data '{"rmsCurrent":120}' from endpoint 1 with groupID 0
`

The second plug can not be controlled, this is shown up in the zigbee-log when I trigger the second plug with it´s button:

`2023-08-02 17:31:37Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 2 with groupID 0
2023-08-02 17:31:37Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 2 with groupID 0`

Big question now - how can I help to integrate this device and it´s missing function?

**Comment by thebang2** [comment]
I managed to make the device at least controllable from zigbee2mqtt, everything works here:


```
const zigbeeHerdsmanConverters = require('zigbee-herdsman-converters');
const fz = zigbeeHerdsmanConverters.fromZigbeeConverters;
const ea = zigbeeHerdsmanConverters.exposes.access;
const e = zigbeeHerdsmanConverters.exposes.presets;

// Define the Zigbee device configuration
const device = {
    fingerprint: [{modelID: 'TS011F', manufacturerName: '_TZ3000_jak16dll'}],
    model: 'TS011F',
    description: 'Smart dual plug (with power monitoring)',
    vendor: 'TuYa',

    // Define the conversion functions for incoming Zigbee messages
    fromZigbee: [
        fz.on_off,
        fz.electrical_measurement,
        fz.metering,
        fz.ignore_basic_report,
        {
            cluster: 'genOnOff',
            type: ['attributeReport', 'readResponse'],
            convert: (model, msg, publish, options, meta) => {
                const endpoint = msg.endpoint.ID;
                const state = {state: msg.data['onOff'] === 1 ? 'ON' : 'OFF'};
                return endpoint === 1 ? {state_1: state.state} : {state_2: state.state};
            },
        },
    ],

    // Define the conversion functions for outgoing Zigbee messages
    toZigbee: [
        {
            key: ['state_1'],
            convertSet: async (entity, key, value, meta) => {
                const endpoint = meta.device.getEndpoint(1);
                await endpoint.command('genOnOff', value === 'ON' ? 'on' : 'off', {}, {transactionSequenceNumber: 0});
            },
            convertGet: async (entity, key, meta) => {
                const endpoint = meta.device.getEndpoint(1);
                await endpoint.read('genOnOff', ['onOff']);
            },
        },
        {
            key: ['state_2'],
            convertSet: async (entity, key, value, meta) => {
                const endpoint = meta.device.getEndpoint(2);
                await endpoint.command('genOnOff', value === 'ON' ? 'on' : 'off', {}, {transactionSequenceNumber: 0});
            },
            convertGet: async (entity, key, meta) => {
                const endpoint = meta.device.getEndpoint(2);
                await endpoint.read('genOnOff', ['onOff']);
            },
        },
    ],

    // Define the configuration function for the device
    configure: async (device, coordinatorEndpoint, logger) => {
        // Endpoint 1: Power measurement and switch
        const endpoint1 = device.getEndpoint(1);
        await endpoint1.bind('genOnOff', coordinatorEndpoint);
        await endpoint1.bind('haElectricalMeasurement', coordinatorEndpoint);
        await endpoint1.bind('seMetering', coordinatorEndpoint);

        // Endpoint 2: Switch
        const endpoint2 = device.getEndpoint(2);
        await endpoint2.bind('genOnOff', coordinatorEndpoint);

        device.save();
    },

    // Define the exposed functionalities for the device
    exposes: [
        e.switch().withEndpoint('1'), // Switch for endpoint 1
        e.switch().withEndpoint('2'), // Switch for endpoint 2
        e.power(),
        e.current(),
        e.voltage().withAccess(ea.STATE)
    ],
};

module.exports = [device];
```

Unfortunately, I can not get the device to be controlled from HomeAssistant, I always get the message:

Debug 2023-08-04 09:00:51Received MQTT message on 'zigbee2mqtt/0xa4c138eeeea9f8c3/1/set' with data 'ON'
No converter available for 'state' ("ON") 

I´m a little bit lost, does anyone have an idea?

**Comment by thebang2** [comment]
Unfortunately, I am not getting any further here. As far as I have found out, the device should be identical to a TS011F_plug_1, but with a second end point:

`
Debug 2023-08-26 20:01:14Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Debug 2023-08-26 20:01:17Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Debug 2023-08-26 20:01:18Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 2 with groupID 0
Debug 2023-08-26 20:01:19Received Zigbee message from '0xa4c138eeeea9f8c3', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 2 with groupID 0
`

Unfortunately, I can't find the converter of the single unit, which is why I can't try to expand it. Can anyone help me here? I don't expect anyone to write the converter for me, I just need a push in the right direction.

**Comment by Koenkk** [comment]
Let's continue in https://github.com/Koenkk/zigbee2mqtt/issues/18326
